Misleading error log message when failing to bind to a port

### Describe the bug

With this setup:
```
application.properties:  quarkus.http.host=macbook
/etc/hosts: 192.168.1.11 macbook macbook-admin ...
```

and 192.168.1.11 being unreachable, Quarkus fails to start and logs this message:

`Caused by: io.quarkus.runtime.QuarkusBindException: Port(s) already bound: 8089: Can't assign requested address`

This is misleading, and a serious time waster when you forgot to adjust /etc/hosts after moving from one network to another.

### Expected behavior

The log message should clearly indicate that the interface doesn't exist.


### How to Reproduce?

Set quarkus.http.host to a dns name that resolves to an unreachable network address.


### Quarkus version or git rev

3.19.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Not relevant

--- a/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
@@ -1,6 +1,5 @@
 package io.quarkus.runtime;
 
-import java.util.List;
 import java.util.Locale;
 import java.util.Set;
 import java.util.concurrent.locks.Condition;
@@ -9,7 +8,6 @@ import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 import java.util.logging.ConsoleHandler;
 import java.util.logging.Handler;
-import java.util.stream.Collectors;
 
 import jakarta.enterprise.context.spi.CreationalContext;
 import jakarta.enterprise.inject.Any;
@@ -170,35 +168,30 @@ public class ApplicationLifecycleManager {
             Throwable rootCause = ExceptionUtil.getRootCause(e);
             if (exitCodeHandler == null) {
                 Logger applicationLogger = Logger.getLogger(Application.class);
-                if (rootCause instanceof QuarkusBindException) {
-                    List<Integer> ports = ((QuarkusBindException) rootCause).getPorts();
-                    if (ports.size() == 1) {
+                if (rootCause instanceof QuarkusBindException qbe) {
+                    String host = qbe.getHost();
+                    int port = qbe.getPort();
+                    if (QuarkusBindException.isKnownHost(host)) {
                         applicationLogger.errorf("Port %d seems to be in use by another process. " +
-                                "Quarkus may already be running or the port is used by another application.", ports.get(0));
-                    } else {
-                        applicationLogger.errorf(
-                                "One or more of the following ports: %s seem to be in use by another process. " +
-                                        "Quarkus may already be running or one of the ports is used by another application.",
-                                ports.stream().map(
-                                        Object::toString).collect(Collectors.joining(",")));
-                    }
-                    if (IS_WINDOWS) {
-                        applicationLogger.warn("Use 'netstat -a -b -n -o' to identify the process occupying the port.");
-                        applicationLogger.warn("You can try to kill it with 'taskkill /PID <pid>' or via the Task Manager.");
-                    } else if (IS_MAC) {
-                        for (Integer port : ports) {
+                                "Quarkus may already be running or the port is used by another application.", port);
+                        if (IS_WINDOWS) {
+                            applicationLogger.warn("Use 'netstat -a -b -n -o' to identify the process occupying the port.");
+                            applicationLogger
+                                    .warn("You can try to kill it with 'taskkill /PID <pid>' or via the Task Manager.");
+                        } else if (IS_MAC) {
                             applicationLogger
                                     .warnf("Use 'netstat -anv | grep %d' to identify the process occupying the port.", port);
-                        }
-                        applicationLogger.warn("You can try to kill it with 'kill -9 <pid>'.");
-                    } else {
-                        for (Integer port : ports) {
+                            applicationLogger.warn("You can try to kill it with 'kill -9 <pid>'.");
+                        } else {
                             applicationLogger
                                     .warnf("Use 'ss -anop | grep %1$d' or 'netstat -anop | grep %1$d' to identify the process occupying the port.",
                                             port);
+                            applicationLogger.warn("You can try to kill it with 'kill -9 <pid>'.");
                         }
-                        applicationLogger.warn("You can try to kill it with 'kill -9 <pid>'.");
+                    } else {
+                        applicationLogger.errorf("Unable to bind to host: %s and port: %d.", host, port);
                     }
+
                 } else if (rootCause instanceof ConfigurationException || rootCause instanceof ConfigValidationException) {
                     System.err.println(rootCause.getMessage());
                 } else if (rootCause instanceof PreventFurtherStepsException

--- a/core/runtime/src/main/java/io/quarkus/runtime/QuarkusBindException.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/QuarkusBindException.java
@@ -1,9 +1,6 @@
 package io.quarkus.runtime;
 
 import java.net.BindException;
-import java.util.Arrays;
-import java.util.List;
-import java.util.stream.Collectors;
 
 /**
  * An exception that is meant to stand in for {@link BindException} and provide information
@@ -11,39 +8,33 @@ import java.util.stream.Collectors;
  */
 public class QuarkusBindException extends BindException {
 
-    private final List<Integer> ports;
+    private final String host;
+    private final int port;
 
-    private static String createMessage(List<Integer> ports) {
-        return "Port(s) already bound: " + ports.stream().map(i -> Integer.toString(i)).collect(Collectors.joining(", "));
+    public QuarkusBindException(String host, int port, BindException e) {
+        super(createMessage(host, port) + ": " + e.getMessage());
+        this.host = host;
+        this.port = port;
     }
 
-    private static void assertPortsNotEmpty(List<Integer> ports) {
-        if (ports.isEmpty()) {
-            throw new IllegalStateException("ports must not be empty");
+    private static String createMessage(String host, int port) {
+        // in all these cases, the only thing that can cause a bind exception is the port being in use
+        if (isKnownHost(host)) {
+            return "Port already bound: " + port;
+        } else {
+            return "Unable to bind to host: " + host + " and port: " + port;
         }
     }
 
-    public QuarkusBindException(Integer... ports) {
-        this(Arrays.asList(ports));
+    public static boolean isKnownHost(String host) {
+        return "localhost".equals(host) || "127.0.0.1".equals(host) || "0.0.0.0".equals(host);
     }
 
-    public QuarkusBindException(List<Integer> ports) {
-        super(createMessage(ports));
-        assertPortsNotEmpty(ports);
-        this.ports = ports;
+    public String getHost() {
+        return host;
     }
 
-    public QuarkusBindException(BindException e, Integer... ports) {
-        this(e, Arrays.asList(ports));
-    }
-
-    public QuarkusBindException(BindException e, List<Integer> ports) {
-        super(createMessage(ports) + ": " + e.getMessage());
-        assertPortsNotEmpty(ports);
-        this.ports = ports;
-    }
-
-    public List<Integer> getPorts() {
-        return ports;
+    public int getPort() {
+        return port;
     }
 }

--- a/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/GrpcServerRecorder.java
+++ b/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/GrpcServerRecorder.java
@@ -380,7 +380,7 @@ public class GrpcServerRecorder {
             VertxServer vertxServer = (VertxServer) server;
             vertxServer.start(ar -> {
                 if (ar.failed()) {
-                    Throwable effectiveCause = getEffectiveThrowable(ar, portToServer);
+                    Throwable effectiveCause = getEffectiveThrowable(ar, configuration.host(), portToServer.getKey());
                     if (effectiveCause instanceof QuarkusBindException) {
                         LOGGER.error("Unable to start the gRPC server");
                     } else {
@@ -467,13 +467,13 @@ public class GrpcServerRecorder {
         return definitions;
     }
 
-    private Throwable getEffectiveThrowable(AsyncResult<Void> ar, Map.Entry<Integer, Server> portToServer) {
+    private Throwable getEffectiveThrowable(AsyncResult<Void> ar, String host, int port) {
         Throwable effectiveCause = ar.cause();
         while (effectiveCause.getCause() != null) {
             effectiveCause = effectiveCause.getCause();
         }
-        if (effectiveCause instanceof BindException) {
-            effectiveCause = new QuarkusBindException(portToServer.getKey());
+        if (effectiveCause instanceof BindException e) {
+            effectiveCause = new QuarkusBindException(host, port, e);
         }
         return effectiveCause;
     }
@@ -716,7 +716,7 @@ public class GrpcServerRecorder {
                 VertxServer server = (VertxServer) grpcServer;
                 server.start(ar -> {
                     if (ar.failed()) {
-                        Throwable effectiveCause = getEffectiveThrowable(ar, portToServer);
+                        Throwable effectiveCause = getEffectiveThrowable(ar, configuration.host(), portToServer.getKey());
                         if (effectiveCause instanceof QuarkusBindException) {
                             LOGGER.error("Unable to start the gRPC server");
                         } else {

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -12,7 +12,6 @@ import java.net.URI;
 import java.net.URISyntaxException;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -853,34 +852,17 @@ public class VertxHttpRecorder {
         // See https://vertx.io/docs/vertx-core/java/#_server_sharing for more information
         AtomicBoolean startEventsFired = new AtomicBoolean();
 
-        vertx.deployVerticle(new Supplier<Verticle>() {
+        vertx.deployVerticle(new Supplier<>() {
             @Override
             public Verticle get() {
                 return new WebDeploymentVerticle(httpMainServerOptions, httpMainSslServerOptions, httpMainDomainSocketOptions,
                         launchMode, insecureRequestStrategy, httpConfig, connectionCount, registry, startEventsFired);
             }
-        }, new DeploymentOptions().setInstances(ioThreads), new Handler<AsyncResult<String>>() {
+        }, new DeploymentOptions().setInstances(ioThreads), new Handler<>() {
             @Override
             public void handle(AsyncResult<String> event) {
                 if (event.failed()) {
-                    Throwable effectiveCause = event.cause();
-                    if (effectiveCause instanceof BindException) {
-                        List<Integer> portsUsed = Collections.emptyList();
-
-                        if ((httpMainSslServerOptions == null) && (httpMainServerOptions != null)) {
-                            portsUsed = List.of(httpMainServerOptions.getPort());
-                        } else if ((insecureRequestStrategy == InsecureRequests.DISABLED)
-                                && (httpMainSslServerOptions != null)) {
-                            portsUsed = List.of(httpMainSslServerOptions.getPort());
-                        } else if ((httpMainSslServerOptions != null)
-                                && (insecureRequestStrategy == InsecureRequests.ENABLED)
-                                && (httpMainServerOptions != null)) {
-                            portsUsed = List.of(httpMainServerOptions.getPort(), httpMainSslServerOptions.getPort());
-                        }
-
-                        effectiveCause = new QuarkusBindException((BindException) effectiveCause, portsUsed);
-                    }
-                    futureResult.completeExceptionally(effectiveCause);
+                    futureResult.completeExceptionally(event.cause());
                 } else {
                     futureResult.complete(event.result());
                 }
@@ -1362,7 +1344,11 @@ public class VertxHttpRecorder {
                 @Override
                 public void handle(AsyncResult<HttpServer> event) {
                     if (event.cause() != null) {
-                        startFuture.fail(event.cause());
+                        if (event.cause() instanceof BindException e) {
+                            startFuture.fail(new QuarkusBindException(options.getHost(), options.getPort(), e));
+                        } else {
+                            startFuture.fail(event.cause());
+                        }
                     } else {
                         // Port may be random, so set the actual port
                         int actualPort = event.result().actualPort();

